bluetooth/src/iosMain/kotlin/BluetoothMonitor.kt

```diff
@@ -17,44 +17,62 @@
 
 package com.splendo.kaluga.bluetooth
 
-import com.splendo.kaluga.base.DefaultServiceMonitor
-import com.splendo.kaluga.base.ServiceMonitor
+import co.touchlab.stately.concurrency.AtomicBoolean
+import co.touchlab.stately.concurrency.AtomicReference
+import co.touchlab.stately.concurrency.value
+import com.splendo.kaluga.base.monitor.DefaultServiceMonitor
+import com.splendo.kaluga.base.monitor.ServiceMonitor
 import platform.CoreBluetooth.CBCentralManager
 import platform.CoreBluetooth.CBCentralManagerDelegateProtocol
 import platform.CoreBluetooth.CBCentralManagerStatePoweredOn
 import platform.darwin.NSObject
 
 actual interface BluetoothMonitor : ServiceMonitor {
     actual class Builder(
-        private val centralManager: CBCentralManager = CBCentralManager()
+        private val centralManagerBuilder: () -> CBCentralManager = { CBCentralManager() }
     ) {
-        actual fun create(): BluetoothMonitor = DefaultBluetoothMonitor(centralManager = centralManager)
+        actual fun create(): BluetoothMonitor = DefaultBluetoothMonitor(centralManagerBuilder = centralManagerBuilder)
     }
 }
 
 class DefaultBluetoothMonitor internal constructor(
-    private val centralManager: CBCentralManager
+    private val centralManagerBuilder: () -> CBCentralManager
 ) : DefaultServiceMonitor(), BluetoothMonitor {
 
     internal class CentralManagerDelegate(
         private val updateEnabledState: () -> Unit
     ) : NSObject(), CBCentralManagerDelegateProtocol {
+
+        val isMonitoring = AtomicBoolean(false)
+
         override fun centralManagerDidUpdateState(central: CBCentralManager) {
-            updateEnabledState()
+            if (isMonitoring.value) {
+                updateEnabledState()
+            }
         }
     }
 
+    private val centralManager = AtomicReference<CBCentralManager?>(null)
+
     private val centralManagerDelegate = CentralManagerDelegate(::updateState)
     override val isServiceEnabled: Boolean
-        get() = centralManager.state == CBCentralManagerStatePoweredOn
+        get() = initializeCentralManagerIfNotInitialized().state == CBCentralManagerStatePoweredOn
+
+    private fun initializeCentralManagerIfNotInitialized(): CBCentralManager = centralManager.value ?: centralManagerBuilder().apply {
+        delegate = centralManagerDelegate
+        centralManager.set(this)
+    }
 
     override fun startMonitoring() {
         super.startMonitoring()
-        centralManager.delegate = centralManagerDelegate
+        initializeCentralManagerIfNotInitialized()
+        updateState()
+        centralManagerDelegate.isMonitoring.value = true
     }
 
     override fun stopMonitoring() {
         super.stopMonitoring()
-        centralManager.delegate = null
+        initializeCentralManagerIfNotInitialized()
```

Comment:
After stop monitoring delegate will be still set and centralManagerDidUpdateState called (but you added `if` I saw)

Reply:
Yeah I can try with just updating the delegate but I think we will run into freezing issues

Reply:
So I just opted to disable actually using the delegate

Reply:
Looks like it works fine, reverted slightly